src/main/kotlin/com/cognifide/gradle/aem/tooling/tail/Parser.kt

```diff
@@ -0,0 +1,57 @@
+package com.cognifide.gradle.aem.tooling.tail
+
+import java.io.BufferedReader
+
+class Parser {
+
+    fun parseLogs(reader: BufferedReader): List<Log> {
+        val firstLogLine = skipIncompleteLog(reader)
+        return parseLogs(reader, firstLogLine)
+    }
+
+    private fun parseLogs(reader: BufferedReader, firstLineOfLog: String?): List<Log> {
+        val (log, firstLineOfNextLog) = readLog(reader, firstLineOfLog)
+        if (log == null) {
+            return emptyList()
+        }
+        if (firstLineOfNextLog == null) {
+            return listOf(log)
+        }
+        return listOf(log) + parseLogs(reader, firstLineOfNextLog)
+    }
+
+    private fun readLog(reader: BufferedReader, firstLogLine: String?): Pair<Log?, String?> {
+        val (completeLogLines, firstLineOfNextLog) = readSubsequentLogLines(reader, firstLogLine)
+        return if (completeLogLines.isNotEmpty()) {
+            Log.create(completeLogLines) to firstLineOfNextLog
+        } else {
+            null to null
+        }
+    }
+
+    private fun readSubsequentLogLines(reader: BufferedReader, logLine: String?): Pair<List<String>, String?> {
+        val followingLine = readLine(reader)
+        return when {
+            logLine == null -> emptyList<String>() to null
```

Comment:
a lot of null checks here... are you sure that you cannot avoid them? generally prefer empty strings, vs nulls etc (some types representing empty states)

Reply:
but in case of performance, probably your version will be better. if it is, keep it.

Reply:
I don't like nulls at all. Happily, Kotlin makes them explicit. Those nulls are here because of multiline parsing. Maybe I could make it nicer but I would wait for an occasion.